keras/src/losses/loss.py

```diff
@@ -24,11 +24,33 @@ def call(self, y_true, y_pred):
     ```
     """
 
+    SPEC_TENSOR = "tensor"
+
     def __init__(self, name=None, reduction="sum_over_batch_size", dtype=None):
         self.name = name or auto_name(self.__class__.__name__)
         self.reduction = standardize_reduction(reduction)
         self.dtype = dtype or backend.floatx()
 
+        # `self.*_spec` can open the possibility to have different signature for
+        # `y_true` and `y_pred`.
+        # By default, `y_true` and `y_pred` will be a sinlge tensor.
+        self.y_true_spec = self.SPEC_TENSOR
+        self.y_pred_spec = self.SPEC_TENSOR
+
+    def set_specs(self, y_true, y_pred):
```

Comment:
It is difficult or even impossible to avoid this function if `y_true` or `y_pred` is nested instead of a plain tensor.

We have no information about the structure of `y_true` and `y_pred` when compiling the losses. This was previously resolved through complicated analysis and a very strong assumption that each loss has single tensors as inputs.

Obviously, the single tensor input format is insufficient for many cases.
(Personally, in an object detection task, I had to stack multiple `y_true` values into a single tensor to make `CompileLoss` works. It took me some time to work around it.)

This solution for nested input structures is the only one I could come up with. It feels overengineered, but it works. Additionally, the `build` part is much cleaner than before.